File: solutions/trees/212-word.py
I'll help you implement a solution for Word Search II in Python. This solution will use a Trie data structure for efficient word searching in the board.

```python
from typing import List

class TrieNode:
    """Node class for Trie data structure"""
    def __init__(self):
        self.children = {}
        self.is_word = False
        self.word = ''

class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        """
        Find all words from the given list that can be formed by adjacent characters on the board.
        
        Args:
            board: 2D list of characters representing the game board
            words: List of words to search for
            
        Returns:
            List of words found on the board
        """
        if not board or not board[0] or not words:
            return []

        # Build Trie
        root = TrieNode()
        for word in words:
            node = root
            for char in word:
                if char not in node.children:
                    node.children[char] = TrieNode()
                node = node.children[char]
            node.is_word = True
            node.word = word

        rows, cols = len(board), len(board[0])
        result = set()

        def dfs(row: int, col: int, node: TrieNode) -> None:
            """
            Depth-first search to find words on the board starting from given position.
            
            Args:
                row: Current row position
                col: Current column position
                node: Current TrieNode in the trie
            """
            # Check if current position is valid
            if (row < 0 or row >= rows or 
                col < 0 or col >= cols or 
                board[row][col] not in node.children):
                return

            char = board[row][col]
            curr_node = node.children[char]

            # If we found a word, add it to results
            if curr_node.is_word:
                result.add(curr_node.word)

            # Mark current cell as visited
            board[row][col] = '#'

            # Search in all four directions
            directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
            for dx, dy in directions:
                new_row, new_col = row + dx, col + dy
                dfs(new_row, new_col, curr_node)

            # Restore the cell
            board[row][col] = char

        # Start DFS from each cell
        for i in range(rows):
            for j in range(cols):
                dfs(i, j, root)

        return list(result)


def test_word_search():
    """Test function with example cases"""
    solution = Solution()
    
    # Test case 1
    board1 = [
        ["o","a","a","n"],
        ["e","t","a","e"],
        ["i","h","k","r"],
        ["i","f","l","v"]
    ]
    words1 = ["oath","pea","eat","rain"]
    assert set(solution.findWords(board1, words1)) == {"eat","oath"}

    # Test case 2
    board2 = [["a","b"],["c","d"]]
    words2 = ["abcb"]
    assert solution.findWords(board2, words2) == []

    print("All test cases passed!")


if __name__ == "__main__":
    test_word_search()
```

This implementation solves the Word Search II problem using the following approach:

1. We create a Trie data structure to store all the words we're looking for. This makes the search process more efficient.

2. The `TrieNode` class represents nodes in our Trie, with:
   - `children`: dictionary of child nodes
   - `is_word`: boolean indicating if this node represents the end of a word
   - `word`: the complete word if this is an end node

3. The main `findWords` method:
   - First builds the Trie from the input words
   - Then performs DFS from each cell on the board
   - Uses a set to store found words (avoiding duplicates)

4. The `dfs` helper function:
   - Performs depth-first search on the board
   - Uses the Trie to guide the search
   - Marks visited cells with '#' to avoid using them again in the same path
   - Restores cells after exploring all possibilities

5. The solution handles edge cases:
   - Empty board or words list
   - Invalid board positions
   - Words that don't exist on the board

6. Includes test cases to verify the implementation

The code follows Python best practices:
- Type hints for better code readability
- Clear documentation and comments
- Proper error handling
- Efficient data structures (Trie, Set)
- Clean code structure with separate concerns

To use this solution, simply call the `findWords` method with a board and list of words as arguments. The test cases demonstrate the expected usage and results.

File: solutions/trees/alternatives/212-word.js
/**
 * 212. Word
 * Medium
 *
 * Word Search II - Find all words in a board of characters Uses Trie data structure for efficient word searching Time Complexity: O(M N 4^L) where M,N are board dimensions and L is max word length Space Complexity: O(K) where K is total length of all words in dictionary
 *
 * <details>
 * <summary><b>🔍 SOLUTION EXPLANATION</b></summary>
 *
 * ### INTUITION:
 * The key insight for solving Word is to understand the core problem pattern
 * and apply the most efficient algorithmic approach.
 *
 * ### APPROACH:
 * 1. Analyze the problem requirements
 * 2. Choose the optimal data structure
 * 3. Implement the solution step by step
 * 4. Handle edge cases appropriately
 *
 * ### WHY THIS WORKS:
 * This approach works because it leverages the fundamental properties of the problem
 * to achieve an efficient solution.
 *
 * ### EXAMPLE WALKTHROUGH:
 * For a typical input, the algorithm processes the data systematically
 * to produce the expected output.
 *
 * </details>
 */

/**
 * Word Search II - Find all words in a board of characters
 * Uses Trie data structure for efficient word searching
 * Time Complexity: O(M * N * 4^L) where M,N are board dimensions and L is max word length
 * Space Complexity: O(K) where K is total length of all words in dictionary
 */

// Trie Node class for building the prefix tree
class TrieNode {
    constructor() {
        this.children = {};
        this.isEndOfWord = false;
        this.word = '';
    }
}

/**
 * @param {character[][]} board
 * @param {string[]} words
 * @return {string[]}
 */
const findWords = function(board, words) {
    if (!board || !board.length || !words || !words.length) {
        return [];
    }

    // Build Trie from words
    const root = buildTrie(words);
    const result = new Set();
    const rows = board.length;
    const cols = board[0].length;

    // Search each cell as starting point
    for (let i = 0; i < rows; i++) {
        for (let j = 0; j < cols; j++) {
            dfs(board, i, j, root, result);
        }
    }

    return Array.from(result);
};

/**
 * Builds a Trie from array of words
 * @param {string[]} words
 * @return {TrieNode}
 */
function buildTrie(words) {
    const root = new TrieNode();

    for (const word of words) {
        let node = root;
        for (const char of word) {
            if (!node.children[char]) {
                node.children[char] = new TrieNode();
            }
            node = node.children[char];
        }
        node.isEndOfWord = true;
        node.word = word;
    }

    return root;
}

/**
 * DFS search through the board starting from position (i,j)
 * @param {character[][]} board
 * @param {number} i
 * @param {number} j
 * @param {TrieNode} node
 * @param {Set} result
 */
function dfs(board, i, j, node, result) {
    // Check boundaries and if character exists in current trie node
    if (i < 0 || i >= board.length || j < 0 || j >= board[0].length || 
        !board[i][j] || !node.children[board[i][j]]) {
        return;
    }

    const char = board[i][j];
    const currentNode = node.children[char];

    // If we found a word, add it to result
    if (currentNode.isEndOfWord) {
        result.add(currentNode.word);
    }

    // Mark cell as visited
    board[i][j] = '#';

    // Search in all four directions
    const directions = [[0, 1], [1, 0], [0, -1], [-1, 0]];
    for (const [dx, dy] of directions) {
        dfs(board, i + dx, j + dy, currentNode, result);
    }

    // Restore the cell
    board[i][j] = char;
}

// Example usage:
const board = [
    ['o','a','a','n'],
    ['e','t','a','e'],
    ['i','h','k','r'],
    ['i','f','l','v']
];
const words = ["oath","pea","eat","rain"];

console.log(findWords(board, words)); // Expected output: ["eat","oath"]

// Export for testing
module.exports = {
    findWords,
    TrieNode
};